src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs

```diff
@@ -233,74 +139,124 @@ public PartialDateTime(DateTimeOffset dateTimeOffset)
+            if (GetIsDateTimePartSpecified(InvalidTimeZoneCapture))
             {
-                utcOffset = TimeSpan.FromMinutes(0);
-            }
-            else if (!string.IsNullOrEmpty(timeZone))
-            {
-                utcOffset = DateTimeOffset.ParseExact(timeZone, "zzz", CultureInfo.InvariantCulture).Offset;
+                throw new FormatException(string.Format(Resources.DateTimeStringIsIncorrectlyFormatted, inputString));
             }
 
-            // In search queries, the time zone information is optional (http://hl7.org/fhir/search.html#date).
-            // The regular expression will not capture the time zone information unless hour and minutes are present.
+            TimeSpan? utcOffset = GetIsDateTimePartSpecified(TimeZoneCapture) ? parsedDateTimeOffset.Offset : (TimeSpan?)null;
+
             // If hour and minutes are specified but time zone information is not, then we will default to UTC
-            // since all dates without time zone is stored in UTC timestamp on the server.
+            // because all dates without time zone information are stored with a UTC timestamp on the server.
             if (hour != null && utcOffset == null)
             {
                 utcOffset = TimeSpan.FromMinutes(0);
             }
 
-            try
+            return new PartialDateTime(year, month, day, hour, minute, second, fraction, utcOffset);
+
+            bool GetIsDateTimePartSpecified(string name)
             {
-                return new PartialDateTime(year, month, day, hour, minute, second, fraction, utcOffset);
+                var stringValue = match.Groups[name]?.Value;
+                return !string.IsNullOrEmpty(stringValue);
             }
-            catch (Exception ex) when (ex is ArgumentException)
+        }
+
+        /// <summary>
+        /// Creates the formats array, which outlines how to parse the input string into a DateTimeOffset object.
+        /// </summary>
+        /// <returns>An array of acceptable DateTimeOffset formats.</returns>
+        /// <remarks>
+        /// The parser serves two different purposes: storing and searching.
+        /// There is a difference between how the date should be parsed for storing and how the date should be parsed for searching.
+        /// For date that should be stored, the time zone information must be present if time is specified.
+        /// From spec: http://hl7.org/fhir/datatypes.html#datetime, "If hours and minutes are specified, a time zone SHALL be populated."
+        /// However, if the date is being parsed for searching, then the time zone information is optional.
+        /// From spec: http://hl7.org/fhir/search.html#date, "you SHOULD provide a time zone if the time part is present."
+        /// As a result, the formats array allows the time zone to be optional.
+        /// </remarks>
+        private static string[] GenerateDateTimeOffsetFormats()
```

Comment:
Does it make sense to generate this array in a method?

Reply:
I think this makes sense. This would be a large array to enumerate every option, and this makes typos less likely and easier to find.